WordPress/src/main/java/org/wordpress/android/ui/posts/PrepublishingAddCategoryFragment.kt

```diff
@@ -0,0 +1,165 @@
+    }
+
+    private fun initBackButton() {
+        back_button.setOnClickListener {
+            viewModel.addCategory(
+                    category_name.text.toString(),
+                    (parent_category.selectedItem as CategoryNode)
+            )
+            viewModel.onBackButtonClicked()
+        }
+    }
+
+    private fun initCloseButton() {
+        close_button.setOnClickListener {
+            viewModel.onBackButtonClicked()
+        }
+    }
+
+    private fun initInputText() {
+        category_name.requestFocus()
+        ActivityUtils.showKeyboard(category_name)
+    }
+
+    private fun initViewModel() {
+        viewModel = ViewModelProviders.of(this, viewModelFactory)
+                .get(PrepublishingAddCategoryViewModel::class.java)
+        startObserving()
+    }
+
+    private fun startObserving() {
+        viewModel.dismissKeyboard.observe(viewLifecycleOwner, Observer { event ->
+            event?.applyIfNotHandled {
+                ActivityUtils.hideKeyboardForced(category_name)
+            }
+        })
+
+        viewModel.navigateBack.observe(viewLifecycleOwner, Observer { event ->
+            event?.applyIfNotHandled {
+                closeListener?.onBackClicked()
+            }
+        })
+
+        viewModel.toolbarTitleUiState.observe(viewLifecycleOwner, Observer { uiString ->
+            toolbar_title.text = uiHelpers.getTextOfUiString(requireContext(), uiString)
+        })
+
+        viewModel.snackbarEvents.observe(viewLifecycleOwner, Observer {
+            it?.applyIfNotHandled {
+                showToast()
+            }
+        })
+
+        viewModel.uiState.observe(viewLifecycleOwner, Observer { uiState ->
+                when (uiState) {
+                    is ContentUiState -> {
+                        loadCategories(uiState.categories)
+                    }
+                }
+            with(uiHelpers) {
+                updateVisibility(close_button, uiState.closeButtonVisible)
+            }
+        })
+
+        val needsRequestLayout = requireArguments().getBoolean(PrepublishingTagsFragment.NEEDS_REQUEST_LAYOUT)
+        val siteModel = requireArguments().getSerializable(WordPress.SITE) as SiteModel
+        viewModel.start(siteModel, !needsRequestLayout)
+    }
+
+    private fun loadCategories(categoryLevels: ArrayList<CategoryNode>) {
+        categoryLevels.add(
+                0, CategoryNode(
+                0, 0,
+                getString(R.string.top_level_category_name)
+        ))
+        val categoryAdapter = ParentCategorySpinnerAdapter(
+                activity,
+                layout.categories_row_parent,
+                categoryLevels
+        )
+        parent_category.adapter = categoryAdapter
+    }
+
+    private fun SnackbarMessageHolder.showToast() {
+        val message = uiHelpers.getTextOfUiString(requireContext(), this.message).toString()
+        ToastUtils.showToast(
+                requireContext(), message,
+                LONG
```

Comment:
I switched to Long because the message doesn't show in the view in which the action was submitted. I wanted the user to have more time to notice if there was an issue.